astropy/table/_dataframes.py

```diff
@@ -0,0 +1,517 @@
+                    nw.when(
+                        nw.from_numpy(
+                            array[n].mask[:, None],  # Reshape into column vector
+                            backend=backend_impl,
+                        )[:, 0]
+                    )
+                    .then(None)
+                    .otherwise(nw.col(n))
+                    .alias(n)
+                )
+                for n in masked_cols
+            ]
+        )
+
+        for n, old_dtype in old_dtypes.items():
+            if old_dtype.is_integer() and not use_nullable_int:
+                raise ValueError(
+                    "Cannot convert masked integer columns to DataFrame without using nullable integers. "
+                    f"Set use_nullable_int=True or remove the offending column: {n}."
+                )
+            elif old_dtype.is_float():
+                df_nw = df_nw.with_columns(nw.col(n).cast(old_dtype).alias(n))
+
+    # Convert to dataframe
+    df_native = df_nw.to_native()
+
+    # Fix pandas-like nullable integers
+    if backend_impl.is_pandas_like() and tbl.has_masked_columns and use_nullable_int:
+        for name in masked_cols:
+            dtype = array[name].dtype
+            if (dtype := array[name].dtype).kind not in INTEGER_DTYPE_KINDS:
+                continue
+
+            df_native[name] = df_native[name].astype(_numpy_to_pandas_dtype(dtype))
+
+    # Pandas-like index
+    if index:
+        df_native.set_index(index, inplace=True, drop=True)
+
+    return df_native
+
+
+def from_df(
+    df: Any, *, index: bool = False, units: Mapping[str, "UnitLike"] | None = None
+) -> "Table":
+    """Create a Table from any narwhals-compatible DataFrame."""
+    from .table import Table
+
+    if not HAS_NARWHALS:
+        raise ModuleNotFoundError(
+            "The narwhals library is required for the generic from_df method. "
+            "If you want to only convert from pandas, use the `from_pandas` method instead."
+        )
+    import narwhals as nw
+
+    # Create output
+    out = OrderedDict()
+
+    # Handle pandas index
+    if index:
+        if not hasattr(df, "index"):
+            raise ValueError(
+                "The input dataframe does not have an index. "
+                "Are you trying to convert a non-pandas dataframe? "
+                "Set `index=False` to avoid this error."
+            )
+        index_name = df.index.name or "index"
+        while index_name in df.columns:
+            index_name = "_" + index_name + "_"
+        df.reset_index(index_name, inplace=True, drop=False)
+
+    # Narwhals layer, must convert to eager
+    df_nw = nw.from_native(df)
+    if isinstance(df_nw, nw.LazyFrame):
+        df_nw = df_nw.collect()
+
+    # Handle units
+    if units is None:
+        units = {}
+    elif not isinstance(units, Mapping):
+        raise TypeError('Expected a Mapping "column-name" -> "unit"')
+    not_found = set(units.keys()) - set(df_nw.columns)
+    if not_found:
```

Comment:
This is the typical use case for walruses. Also, `set(dict.keys())` should be equivalent to `set(dict)`, which hopefully makes it easier on the eye. Finally, we can use the explicit name of the set operation we need instead of converting both collections and using a builtin operator (which many readers might not understand without looking up)
```suggestion
    if (not_found := set(units).difference(df_nw.columns)):
```